File: project24.py
import numpy as np
import matplotlib.pyplot as plt
import math
from coeffs import *
#Computaion part
#defining the direction vector of the normal of the line
P=np.array([1,1])
print("P=",P)
#finding the direction  vector of the line
m=omat@P
print("m=",m)
#finding the point of intersection
s=np.array(([-1,1],[1,1]))
#intersection1
k1=np.array([0,5])
t1=np.linalg.inv(s)@k1
print("t1=",t1)
#intersection2
k2=np.array([0,4])
t2=np.linalg.inv(s)@k2
print("t2=",t2)
#intersection3
k3=np.array([0,3])
t3=np.linalg.inv(s)@k3
print("t3=",t3)
#intersection4
k4=np.array([0,2])
t4=np.linalg.inv(s)@k4
print("t4=",t4)
#intersection5
k5=np.array([0,1])
t5=np.linalg.inv(s)@k5
print("t5=",t5)
#intersection6
k6=np.array([0,0])
t6=np.linalg.inv(s)@k6
print("t6=",t6)
#Finding their distances of line from centre
#distance from origin to line1 perpendicularly
d1=np.linalg.norm(t1)
print("d1=",d1)
#distance from origin to line2 perpendicularly
d2=np.linalg.norm(t2)
print("d2=",d2)
#distance from origin to line3 perpendicularly
d3=np.linalg.norm(t3)
print("d3=",d3)
#distance from origin to line4 perpendicularly
d4=np.linalg.norm(t4)
print("d4=",d4)
#distance from origin to line5 perpendicularly
d5=np.linalg.norm(t5)
print("d5=",d5)
#distance from origin to line6 perpendicularly
d6=np.linalg.norm(t6)
print("d6=",d6)
#Finding the length of the intercepts
r=4
#intercept1
l1=2*np.sqrt((4*4)-(d1*d1))
print("l1=",l1)
#intercept2
l2=2*np.sqrt((4*4)-(d2*d2))
print("l2=",l2)
#intercept3
l3=2*np.sqrt((4*4)-(d3*d3))
print("l3=",l3)
#intercept4
l4=2*np.sqrt((4*4)-(d4*d4))
print("l4=",l4)
#intercept5
l5=2*np.sqrt((4*4)-(d5*d5))
print("l5=",l5)
#intercept6
l6=2*np.sqrt((4*4)-(d6*d6))
print("l6=",l6)
#Finding the sum of the squares of the intercepts
S=(l1*l1+l2*l2+l3*l3+l4*l4+l5*l5)*2+l6*l6
print("S=",S)
#data for  circle
O=np.array([0,0])
n=np.array([1,1])
r=4
c=O
len = 1000
theta = np.linspace(0,2*np.pi,len)
x_circ = np.zeros((2,len))
x_circ[0,:] = r*np.cos(theta)
x_circ[1,:] = r*np.sin(theta)
x_circ = (x_circ.T + c).T



#data for lines
A1=np.array([4,-2])
B1=np.array([-4,6])
A2=np.array([4,-3])
B2=np.array([-4,5])
A3=np.array([2*np.sqrt(2),-2*np.sqrt(2)])
B3=np.array([-2*np.sqrt(2),2*np.sqrt(2)])
A4=np.array([-4,3])
B4=np.array([4,-5])
A5=np.array([-4,2])
B5=np.array([4,-6])
A6=np.array([-4,7])
B6=np.array([4,-1])
A7=np.array([4,-7])
B7=np.array([-4,1])
A8=np.array([4,0])
B8=np.array([0,4])
A9=np.array([-4,0])
B9=np.array([0,-4])
A10=np.array([5,0])
B10=np.array([0,5])
A11=np.array([-5,0])
B11=np.array([0,-5])


#plotting lines
x_A1B1 = line_gen(A1,B1)
plt.plot(x_A1B1[0,:],x_A1B1[1,:])
x_A2B2 = line_gen(A2,B2)
plt.plot(x_A2B2[0,:],x_A2B2[1,:])
x_A3B3 = line_gen(A3,B3)
plt.plot(x_A3B3[0,:],x_A3B3[1,:])
x_A4B4 = line_gen(A4,B4)
plt.plot(x_A4B4[0,:],x_A4B4[1,:])
x_A5B5 = line_gen(A5,B5)
plt.plot(x_A5B5[0,:],x_A5B5[1,:])
x_A6B6 = line_gen(A6,B6)
plt.plot(x_A6B6[0,:],x_A6B6[1,:])
x_A7B7 = line_gen(A7,B7)
plt.plot(x_A7B7[0,:],x_A7B7[1,:])
x_A8B8 = line_gen(A8,B8)
plt.plot(x_A8B8[0,:],x_A8B8[1,:])
x_A9B9 = line_gen(A9,B9)
plt.plot(x_A9B9[0,:],x_A9B9[1,:])
x_A10B10 = line_gen(A10,B10)
plt.plot(x_A10B10[0,:],x_A10B10[1,:])
x_A11B11 = line_gen(A11,B11)
plt.plot(x_A11B11[0,:],x_A11B11[1,:])

#plotting circle
plt.plot(x_circ[0,:],x_circ[1,:],label='$C_1$')

plt.axis('equal')
plt.xlabel('$x$');plt.ylabel('$y$')
plt.plot(O[0], O[1], 'o')
plt.text(O[0] * (1 + 0.1), O[1] * (1 - 0.1) , 'O')
plt.legend(loc='best');plt.grid()
plt.show()

File: project24.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "coeffs.h"


int main(void)
{

int len=100;
//For finding the data for plotting circle
double** theta = linspace(0,2*M_PI,len);
double r = 4;
double** x = createMat(1,len);
double** y = createMat(1,len);

for (int i=0; i<len; i++){
	x[0][i] = r*cos(*theta[i]);
	y[0][i] = r*sin(*theta[i]);
 
}


savetxt(x,"x.dat",1,len);
savetxt(y,"y.dat",1,len);

//Defining the direction vector of the normal of the given lines
double**P=createMat( 2, 1);
P[0][0]=1;P[1][0]=1;
printf("P=\n");
print(P,2,1);
//Defining the omat matrix
double**omat=createMat( 2, 2);
omat[0][0]=0;omat[0][1]=1;
omat[1][0]=-1;omat[1][1]=0;
printf("omat=\n");
print(omat,2,2);
//finding the direction vector of the given line
double **m=matmul(omat,P,2,2,1);
printf("m=\n");
print(m ,2,1);
//Finding the point of intersections
double**s=createMat( 2, 2);
s[0][0]=-1;s[0][1]=1;
s[1][0]=1;s[1][1]=1;
double **j=linalg_inv( s, 2);
double**k1=createMat( 2, 1);
k1[0][0]=0;k1[1][0]=5;
double **T1=matmul(j,k1,2,2,1);
printf("T1=\n");
print(T1,2,1);
double**k2=createMat( 2, 1);
k2[0][0]=0;k2[1][0]=4;
double **T2=matmul(j,k2,2,2,1);
printf("T2=\n");
print(T2,2,1);
double**k3=createMat( 2, 1);
k3[0][0]=0;k3[1][0]=3;
double **T3=matmul(j,k3,2,2,1);
printf("T3=\n");
print(T3,2,1);
double**k4=createMat( 2, 1);
k4[0][0]=0;k4[1][0]=2;
double **T4=matmul(j,k4,2,2,1);
printf("T4=\n");
print(T4,2,1);
double**k5=createMat( 2, 1);
k5[0][0]=0;k5[1][0]=1;
double **T5=matmul(j,k5,2,2,1);
printf("T5=\n");
print(T5,2,1);
double**k6=createMat( 2, 1);
k6[0][0]=0;k6[1][0]=0;
double **T6=matmul(j,k6,2,2,1);
printf("T6=\n");
print(T6,2,1);
//Finding their distances of line from centre
double d1=linalg_norm(T1,2);
printf("d1=");printf("%lf\n",d1);
double d2=linalg_norm(T2,2);
printf("d2=");printf("%lf\n",d2);
double d3=linalg_norm(T3,2);
printf("d3=");printf("%lf\n",d3);
double d4=linalg_norm(T4,2);
printf("d4=");printf("%lf\n",d4);
double d5=linalg_norm(T5,2);
printf("d5=");printf("%lf\n",d5);
double d6=linalg_norm(T6,2);
printf("d6=");printf("%lf\n",d6);
//Finding the length of the intercepts
double l1=2*sqrt(4*4-d1*d1);
printf("l1=");printf("%lf\n",l1);
double l2=2*sqrt(4*4-d2*d2);
printf("l2=");printf("%lf\n",l2);
double l3=2*sqrt(4*4-d3*d3);
printf("l3=");printf("%lf\n",l3);
double l4=2*sqrt(4*4-d4*d4);
printf("l4=");printf("%lf\n",l4);
double l5=2*sqrt(4*4-d5*d5);
printf("l5=");printf("%lf\n",l5);
double l6=2*sqrt(4*4-d6*d6);
printf("l6=");printf("%lf\n",l6);
//Finding the sum of the squares of the intercepts
double S=(l1*l1+l2*l2+l3*l3+l4*l4+l5*l5)*2+l6*l6;
printf("S=");printf("%lf\n",S);
//saving data for lines
//line1
double**A1=createMat( 2, 1);
A1[0][0]=4;A1[1][0]=0;
double**B1=createMat( 2, 1);
B1[0][0]=0;B1[1][0]=4;
double**A1B1=line_gen(A1,B1);
savetxt(A1,"A1.dat",2,1);
savetxt(B1,"B1.dat",2,1);
savetxt(A1B1,"A1B1.dat",2,len);
//line2
double**A2=createMat( 2, 1);
A2[0][0]=5;A2[1][0]=0;
double**B2=createMat( 2, 1);
B2[0][0]=0;B2[1][0]=5;
double**A2B2=line_gen(A2,B2);
savetxt(A2,"A2.dat",2,1);
savetxt(B2,"B2.dat",2,1);
savetxt(A2B2,"A2B2.dat",2,len);
//line3
double**A3=createMat( 2, 1);
A3[0][0]=-5;A3[1][0]=0;
double**B3=createMat( 2, 1);
B3[0][0]=0;B3[1][0]=-5;
double**A3B3=line_gen(A3,B3);
savetxt(A3,"A3.dat",2,1);
savetxt(B3,"B3.dat",2,1);
savetxt(A3B3,"A3B3.dat",2,len);
//line4
double**A4=createMat( 2, 1);
A4[0][0]=-4;A4[1][0]=0;
double**B4=createMat( 2, 1);
B4[0][0]=0;B4[1][0]=-4;
double**A4B4=line_gen(A4,B4);
savetxt(A4,"A4.dat",2,1);
savetxt(B4,"B4.dat",2,1);
savetxt(A4B4,"A4B4.dat",2,len);
//line5
double**A5=createMat( 2, 1);
A5[0][0]=4;A5[1][0]=-3;
double**B5=createMat( 2, 1);
B5[0][0]=-4;B5[1][0]=5;
double**A5B5=line_gen(A5,B5);
savetxt(A5,"A5.dat",2,1);
savetxt(B5,"B5.dat",2,1);
savetxt(A5B5,"A5B5.dat",2,len);
//line6
double**A6=createMat( 2, 1);
A6[0][0]=-4;A6[1][0]=3;
double**B6=createMat( 2, 1);
B6[0][0]=4;B6[1][0]=-5;
double**A6B6=line_gen(A6,B6);
savetxt(A6,"A6.dat",2,1);
savetxt(B6,"B6.dat",2,1);
savetxt(A6B6,"A6B6.dat",2,len);
//line7
double**A7=createMat( 2, 1);
A7[0][0]=2*sqrt(2);A7[1][0]=-2*sqrt(2);
double**B7=createMat( 2, 1);
B7[0][0]=-2*sqrt(2);B7[1][0]=2*sqrt(2);
double**A7B7=line_gen(A7,B7);
savetxt(A7,"A7.dat",2,1);
savetxt(B7,"B7.dat",2,1);
savetxt(A7B7,"A7B7.dat",2,len);
//line8
double**A8=createMat( 2, 1);
A8[0][0]=-4;A8[1][0]=2;
double**B8=createMat( 2, 1);
B8[0][0]=4;B8[1][0]=-6;
double**A8B8=line_gen(A8,B8);
savetxt(A8,"A8.dat",2,1);
savetxt(B8,"B8.dat",2,1);
savetxt(A8B8,"A8B8.dat",2,len);
//line9
double**A9=createMat( 2, 1);
A9[0][0]=4;A9[1][0]=-2;
double**B9=createMat( 2, 1);
B9[0][0]=-4;B9[1][0]=6;
double**A9B9=line_gen(A9,B9);
savetxt(A9,"A9.dat",2,1);
savetxt(B9,"B9.dat",2,1);
savetxt(A9B9,"A9B9.dat",2,len);
//line10
double**A10=createMat( 2, 1);
A10[0][0]=-4;A10[1][0]=7;
double**B10=createMat( 2, 1);
B10[0][0]=4;B10[1][0]=-1;
double**A10B10=line_gen(A10,B10);
savetxt(A10,"A10.dat",2,1);
savetxt(B10,"B10.dat",2,1);
savetxt(A10B10,"A10B10.dat",2,len);
//line11
double**A11=createMat( 2, 1);
A11[0][0]=4;A11[1][0]=-7;
double**B11=createMat( 2, 1);
B11[0][0]=-4;B11[1][0]=1;
double**A11B11=line_gen(A11,B11);
savetxt(A11,"A11.dat",2,1);
savetxt(B11,"B11.dat",2,1);
savetxt(A11B11,"A11B11.dat",2,len);

return 0;
}
